shell/platform/embedder/tests/embedder_a11y_unittests.cc

```diff
@@ -177,98 +187,52 @@ TEST_F(EmbedderA11yTest, A11yTreeIsConsistent) {
   auto engine = builder.LaunchEngine();
   ASSERT_TRUE(engine.is_valid());
 
-  // Wait for initial NotifySemanticsEnabled(false).
-  fml::AutoResetWaitableEvent notify_semantics_enabled_latch;
-  notify_semantics_enabled_callback = [&](Dart_NativeArguments args) {
-    Dart_Handle exception = nullptr;
-    bool enabled =
-        ::tonic::DartConverter<bool>::FromArguments(args, 0, exception);
-    ASSERT_FALSE(enabled);
-    notify_semantics_enabled_latch.Signal();
-  };
-  notify_semantics_enabled_latch.Wait();
-
-  // Prepare to NotifyAccessibilityFeatures call
-  fml::AutoResetWaitableEvent notify_features_latch;
-  notify_accessibility_features_callback = [&](Dart_NativeArguments args) {
-    Dart_Handle exception = nullptr;
-    bool enabled =
-        ::tonic::DartConverter<bool>::FromArguments(args, 0, exception);
-    ASSERT_FALSE(enabled);
-    notify_features_latch.Signal();
-  };
-
-  // Enable semantics. Wait for NotifySemanticsEnabled(true).
-  fml::AutoResetWaitableEvent notify_semantics_enabled_latch_2;
-  notify_semantics_enabled_callback = [&](Dart_NativeArguments args) {
-    Dart_Handle exception = nullptr;
-    bool enabled =
-        ::tonic::DartConverter<bool>::FromArguments(args, 0, exception);
-    ASSERT_TRUE(enabled);
-    notify_semantics_enabled_latch_2.Signal();
-  };
+  // 1: Wait for initial notifySemanticsEnabled(false).
+  semantics_enabled = true;
+  latch.Wait();
+  ASSERT_FALSE(semantics_enabled);
+
+  // 2: Enable semantics. Wait for notifySemanticsEnabled(true).
+  semantics_enabled = false;
   auto result = FlutterEngineUpdateSemanticsEnabled(engine.get(), true);
   ASSERT_EQ(result, FlutterEngineResult::kSuccess);
-  notify_semantics_enabled_latch_2.Wait();
-
-  // Wait for initial accessibility features (reduce_motion == false)
-  notify_features_latch.Wait();
-
-  // Set accessibility features: (reduce_motion == true)
-  fml::AutoResetWaitableEvent notify_features_latch_2;
-  notify_accessibility_features_callback = [&](Dart_NativeArguments args) {
-    Dart_Handle exception = nullptr;
-    bool enabled =
-        ::tonic::DartConverter<bool>::FromArguments(args, 0, exception);
-    ASSERT_TRUE(enabled);
-    notify_features_latch_2.Signal();
-  };
+  latch.Wait();
+  ASSERT_TRUE(semantics_enabled);
+
+  // 3: Wait for notifyAccessibilityFeatures (reduce_motion == false)
+  feature_enabled = true;
+  latch.Wait();
+  ASSERT_FALSE(feature_enabled);
+
+  // 4: Wait for notifyAccessibilityFeatures (reduce_motion == true)
+  feature_enabled = false;
   result = FlutterEngineUpdateAccessibilityFeatures(
       engine.get(), kFlutterAccessibilityFeatureReduceMotion);
   ASSERT_EQ(result, FlutterEngineResult::kSuccess);
-  notify_features_latch_2.Wait();
+  latch.Wait();
+  ASSERT_TRUE(feature_enabled);
 
-  // Wait for UpdateSemantics callback on platform (current) thread.
-  signal_native_latch.Wait();
+  // 5: Wait for UpdateSemantics callback on platform (current) thread.
+  latch.Wait();
```

Comment:
I don't think we can chain multiple `Wait`s on the same latch. This previously caused race conditions that were fixed by separating latches: https://github.com/flutter/engine/pull/37488

Reply:
Before merging this, please test this with a very high `--repeat` value

Reply:
Yep - step 5 requires two latches, and in fact we really want to do testing on the UI thread callback. I've refactored to prototype an alternative.